**Loan default dataset analysis**

The loan default dataset contains information about borrowers who applied for loans, along with details about their financial status, loan characteristics, and repayment behavior. The goal is to perform an analysis and create a report that will assist loan providers in deciding who should be approved for a loan and who should not.

In [206]:
#Importing Libraries 

import pandas as pd
from datetime import datetime

In [207]:
#Importing the datset

loan_dataset = pd.read_csv("Loan_default.csv")

In [208]:
#Dataset preview

loan_dataset.head()

LoanID  Age  Income  LoanAmount  CreditScore  MonthsEmployed  \
0  I38PQUQS96   56   85994       50587          520              80   
1  HPSK72WA7R   69   50432      124440          458              15   
2  C1OZ6DPJ8Y   46   84208      129188          451              26   
3  V2KKSFM3UN   32   31713       44799          743               0   
4  EY08JDHTZP   60   20437        9139          633               8   

   NumCreditLines  InterestRate  LoanTerm  DTIRatio    Education  \
0               4         15.23        36      0.44   Bachelor's   
1               1          4.81        60      0.68     Master's   
2               3         21.17        24      0.31     Master's   
3               3          7.07        24      0.23  High School   
4               4          6.51        48      0.73   Bachelor's   

  EmploymentType MaritalStatus HasMortgage HasDependents LoanPurpose  \
0      Full-time      Divorced         Yes           Yes       Other   
1      Full-time       Married          No            No       Other   
2     Unemployed      Divorced         Yes           Yes        Auto   
3      Full-time       Married          No            No    Business   
4     Unemployed      Divorced          No           Yes        Auto   

  HasCoSigner  Default Loan Date (DD/MM/YYYY)  
0         Yes        0             10/15/2018  
1         Yes        0              3/25/2016  
2          No        1             11/11/2013  
3          No        0              6/22/2017  
4          No        0               6/9/2014

In [209]:
#Finding the number of records

loan_dataset.shape

(255347, 19)

The dataset has 255347 borrowers with 19 metrics collected about them.

**Features**

In [210]:
loan_dataset.columns

Index(['LoanID', 'Age', 'Income', 'LoanAmount', 'CreditScore',
       'MonthsEmployed', 'NumCreditLines', 'InterestRate', 'LoanTerm',
       'DTIRatio', 'Education', 'EmploymentType', 'MaritalStatus',
       'HasMortgage', 'HasDependents', 'LoanPurpose', 'HasCoSigner', 'Default',
       'Loan Date (DD/MM/YYYY)'],
      dtype='object')

**Data dictionary**

**LoanID**: A unique identifier for each loan in the datset

**Age**: The borrower's age at the time the loan was issued

**Income**: The borrower's annual income

**LoanAmount**: The total amount of the loan that the borrower is
                requesting or has been approved for

**CreditScore**: A numerical representation of the borrower's creditworthiness,
                typically ranging from 300 to 850. A higher credit score indicates
                the borrower is more likely to repay the debt

**MonthsEmployed**: The number of months the borrower has been employed their
                    current job or with their current employer

**NumCreditLines**: The total number of active credit lines(e.g credit cards,loans)
                    the borrower has at the time of the loan

**IntrestRate**: The annual percentage rate(APR) charged for borrowing the loan amount,
                 usually expressed as a percentage

**LoanTerm**: The length of time(in months) over which the loan is to be paid

**DTIRatio**: The DTI ratio, which measures the borrower's debt payments relative to their
              income. A higher ration can indicate greater financial stress

**Education**: The highest level of education the borrower has completed(e.g High school,
                Bachelor,Masters)

**EmploymentType**: The type of employment the borrower is engaged in(E.g Full-Time,Part-Time,
                    Self-employed)

**Marital status**: The marital status of the borrowe(single,married or divorced)

**HasMortgage**: An indicator(Yes/No) that shows whether the borrower has an existing mortgage
                 on a property

**HasDependents**: An indicator(Yes/No) that shows whether the borrower has dependents(children,
                    other family members) to support

**LoanPurpose**: The primary reason for taking the loan(Home purchase,Debt consolidation,Education)

**HasCoSigner**: An indicator (Yes/No) that shows whether the borrower has a co-signer for the loan
                (someone who agrees to take responsibility if the borower defaults)

**Default**: An indicator(Yes/No) that shows whether the borrower defaulted on the loan or failed
            to make timely payments.

**Loan Date(DD/MM/YY)**: The date the loan was issued or originated

In [211]:
#Inspecting columns

loan_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 255347 entries, 0 to 255346
Data columns (total 19 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   LoanID                  255347 non-null  object 
 1   Age                     255347 non-null  int64  
 2   Income                  255347 non-null  int64  
 3   LoanAmount              255347 non-null  int64  
 4   CreditScore             255347 non-null  int64  
 5   MonthsEmployed          255347 non-null  int64  
 6   NumCreditLines          255347 non-null  int64  
 7   InterestRate            255347 non-null  float64
 8   LoanTerm                255347 non-null  int64  
 9   DTIRatio                255347 non-null  float64
 10  Education               255347 non-null  object 
 11  EmploymentType          255347 non-null  object 
 12  MaritalStatus           255347 non-null  object 
 13  HasMortgage             255347 non-null  object 
 14  HasDependents       

The year column shows the expected format should be (DD/MM/YYYY) this is however not the case.

In [212]:
loan_dataset["Loan Date (DD/MM/YYYY)"]

0         10/15/2018
1          3/25/2016
2         11/11/2013
3          6/22/2017
4           6/9/2014
             ...    
255342     10/7/2017
255343      1/1/2016
255344      3/5/2017
255345      4/5/2013
255346     1/27/2015
Name: Loan Date (DD/MM/YYYY), Length: 255347, dtype: object

In [213]:
#Changing the date column to (DD/MM/YYYY)

#Converting the date column to the appropiate format
loan_dataset["Loan Date"] = pd.to_datetime(loan_dataset["Loan Date (DD/MM/YYYY)"],format="%m/%d/%Y").dt.strftime("%d/%m/%Y")

In [214]:
# Confirming the date column has been formated appropiately
loan_dataset["Loan Date"]

0         15/10/2018
1         25/03/2016
2         11/11/2013
3         22/06/2017
4         09/06/2014
             ...    
255342    07/10/2017
255343    01/01/2016
255344    05/03/2017
255345    05/04/2013
255346    27/01/2015
Name: Loan Date, Length: 255347, dtype: object

In [215]:
loan_dataset.shape

(255347, 20)

In [216]:
#Dropping the original date column

loan_dataset.drop(columns=["Loan Date (DD/MM/YYYY)"], inplace=True)

In [217]:
loan_dataset.shape

(255347, 19)

**Income colum analysis**

In [218]:
loan_dataset["Income"]

0         85994
1         50432
2         84208
3         31713
4         20437
          ...  
255342    37979
255343    51953
255344    84820
255345    85109
255346    22418
Name: Income, Length: 255347, dtype: int64

In [219]:
minimum_income = loan_dataset["Income"].min()
maximum_income = loan_dataset["Income"].max()

print(maximum_income)
print(minimum_income)

149999
15000


The levels of income ranges frrom **15000** to **149999**.

In [220]:


#categorizing the level of income in 4
bins = [14999, 59999, 104999, 149999]
labels = ['Low Income', 'Medium Income', 'High Income']
loan_dataset['IncomeCategory'] = pd.cut(loan_dataset['Income'], bins=bins, labels=labels)

#Group by IncomeCategory and EmploymentType, then sum loan amounts
grouped_loans = loan_dataset.groupby(['IncomeCategory', 'EmploymentType'])['LoanAmount'].sum().reset_index()
grouped_loans.rename(columns={'LoanAmount': 'TotalLoanAmount'}, inplace=True)

# Step 4: Export to CSV
grouped_loans.to_csv("loan_sum_by_income_and__employment.csv", index=False)


C:\Users\ianob\AppData\Local\Temp\ipykernel_12928\437166228.py:7: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  grouped_loans = loan_dataset.groupby(['IncomeCategory', 'EmploymentType'])['LoanAmount'].sum().reset_index()


In [221]:
grouped_loans

IncomeCategory EmploymentType  TotalLoanAmount
0      Low Income      Full-time       2669960076
1      Low Income      Part-time       2731764200
2      Low Income  Self-employed       2690887374
3      Low Income     Unemployed       2752711341
4   Medium Income      Full-time       2722590175
5   Medium Income      Part-time       2728551426
6   Medium Income  Self-employed       2709549564
7   Medium Income     Unemployed       2714816973
8     High Income      Full-time       2721671577
9     High Income      Part-time       2709285442
10    High Income  Self-employed       2718045487
11    High Income     Unemployed       2707046937

**Credit score analysis**

In [222]:
loan_dataset["CreditScore"]

0         520
1         458
2         451
3         743
4         633
         ... 
255342    541
255343    511
255344    597
255345    809
255346    636
Name: CreditScore, Length: 255347, dtype: int64

In [223]:
max_credit_score = loan_dataset["CreditScore"].max()
minimum_credit_score = loan_dataset["CreditScore"].min()

In [224]:
print(max_credit_score)
print(minimum_credit_score)

849
300


Credit scores, ranging from 300 to 849, are grouped into categories from Very Low to High. The analysis below evaluates the median loan amount within each credit score category

In [225]:
bins = [200,362,525,688,849]
labels = [ "Very Low","Low","Medium","High"]

#Categorizing credit scores into groups
loan_dataset["ScoreCategory"] = pd.cut(loan_dataset["CreditScore"],bins=bins,labels=labels)

#Calculating the median score by category
median_loan_by_score = loan_dataset.groupby("ScoreCategory")["LoanAmount"].median().reset_index()

#Exporting to CSV
median_loan_by_score.to_csv("Median_loan_by_score.csv",index=False)



C:\Users\ianob\AppData\Local\Temp\ipykernel_12928\2294234313.py:8: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  median_loan_by_score = loan_dataset.groupby("ScoreCategory")["LoanAmount"].median().reset_index()


In [226]:
median_loan_by_score

ScoreCategory  LoanAmount
0      Very Low    127227.0
1           Low    128019.0
2        Medium    127795.0
3          High    127015.0

**Analysis of Loan amount by Year**

In [227]:
#Extracting the year column from the datset
loan_dataset["LoanYear"] = pd.to_datetime(loan_dataset["Loan Date"]).dt.year

#Aggregating the loan amount per year
loan_amount_by_year = loan_dataset.groupby("LoanYear")["LoanAmount"].sum().reset_index()

loan_amount_by_year.to_csv("loan_amount_by_year.csv",index=False)

C:\Users\ianob\AppData\Local\Temp\ipykernel_12928\898462524.py:2: UserWarning: Parsing dates in %d/%m/%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  loan_dataset["LoanYear"] = pd.to_datetime(loan_dataset["Loan Date"]).dt.year


In [228]:
loan_amount_by_year

LoanYear  LoanAmount
0      2013  5456934410
1      2014  5373404266
2      2015  5443296583
3      2016  5442736635
4      2017  5383718289
5      2018  5476790389

In [229]:
loan_amount_by_year.max()

LoanYear            2018
LoanAmount    5476790389
dtype: int64

**Analyzing the loan amounts issued by purpose**

In [230]:
loan_dataset["LoanPurpose"].value_counts()

LoanPurpose
Business     51298
Home         51286
Education    51005
Other        50914
Auto         50844
Name: count, dtype: int64

In [231]:
loan_amount_by_purpose = loan_dataset.groupby("LoanPurpose")["LoanAmount"].sum().reset_index()
loan_amount_by_purpose.to_csv("loan_amount_by_purpose.csv", index=False)


In [232]:
loan_amount_by_purpose

LoanPurpose  LoanAmount
0        Auto  6500807511
1    Business  6522120439
2   Education  6510575194
3        Home  6545241527
4       Other  6498135901

**Average loan amount issued by age groups**

In [233]:
#Checking the age column
loan_dataset["Age"]

0         56
1         69
2         46
3         32
4         60
          ..
255342    19
255343    32
255344    56
255345    42
255346    62
Name: Age, Length: 255347, dtype: int64

In [234]:
maximum_age = loan_dataset["Age"].max()
minimum_age = loan_dataset["Age"].min()
print(maximum_age)
print(minimum_age)

69
18


In [235]:
#Creating age groups

bins = [18,25,45,60,70]

labels = ["Teen","Adults","Middle Age Aduls","Senior Citizen"]

#Creating the age group column

loan_dataset["AgeGroup"] = pd.cut(loan_dataset["Age"],bins=bins,labels=labels)

age_group_summary = loan_dataset.groupby("AgeGroup")["LoanAmount"].agg(AverageLoanAmount="mean").reset_index()

age_group_summary["AverageLoanAmount"] = age_group_summary["AverageLoanAmount"].round(2)

age_group_summary.to_csv("age_group_loan.csv",index=False)



C:\Users\ianob\AppData\Local\Temp\ipykernel_12928\3067527072.py:11: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  age_group_summary = loan_dataset.groupby("AgeGroup")["LoanAmount"].agg(AverageLoanAmount="mean").reset_index()


**Analyzing the loan amount amount by type of employment**

In [236]:
loan_amount_by_employment = loan_dataset.groupby("EmploymentType")["LoanAmount"].sum().reset_index()
loan_amount_by_employment.to_csv("loan_amount_by_employment.csv", index=False)


In [237]:
loan_amount_by_employment

EmploymentType  LoanAmount
0      Full-time  8114221828
1      Part-time  8169601068
2  Self-employed  8118482425
3     Unemployed  8174575251

**Loan amount by education levels**

In [238]:
# Aggregate loan amount by education level
loan_amount_by_education = loan_dataset.groupby("Education")["LoanAmount"].sum().reset_index()

# Export to CSV
loan_amount_by_education.to_csv("loan_amount_by_education.csv", index=False)


In [239]:
loan_amount_by_education

Education  LoanAmount
0   Bachelor's  8199625421
1  High School  8140838130
2     Master's  8114883790
3          PhD  8121533231

**Default rate analysis**

In [240]:
loan_dataset["Default"]

0         0
1         0
2         1
3         0
4         0
         ..
255342    0
255343    1
255344    0
255345    0
255346    0
Name: Default, Length: 255347, dtype: int64

Total number of defaulted loans

In [241]:
# Count default values
default_counts = loan_dataset["Default"].value_counts().reset_index()
default_counts.columns = ["Default", "Count"]

# Map 0 → "No", 1 → "Yes"
default_counts["Default"] = default_counts["Default"].map({0: "No", 1: "Yes"})

# Export to CSV
default_counts.to_csv("default_summary.csv", index=False)


In [242]:
default_counts

Default   Count
0      No  225694
1     Yes   29653

**Default rate by Year**

In [243]:
# Extract year from loan date
loan_dataset["LoanYear"] = pd.to_datetime(loan_dataset["Loan Date"]).dt.year

# Group by year and calculate totals and default rate
default_by_year = loan_dataset.groupby("LoanYear")["Default"].agg(
    TotalLoans="count",
    NumberOfDefaults="sum"
).reset_index()

# Add default rate column
default_by_year["DefaultRate (%)"] = round((
    default_by_year["NumberOfDefaults"] / default_by_year["TotalLoans"]
) * 100,2)
default_by_year.to_csv("year_default_rate_summary.csv", index=False)

C:\Users\ianob\AppData\Local\Temp\ipykernel_12928\4125084191.py:2: UserWarning: Parsing dates in %d/%m/%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  loan_dataset["LoanYear"] = pd.to_datetime(loan_dataset["Loan Date"]).dt.year


In [244]:
default_by_year

LoanYear  TotalLoans  NumberOfDefaults  DefaultRate (%)
0      2013       42785              4973            11.62
1      2014       42122              4845            11.50
2      2015       42521              4976            11.70
3      2016       42705              5017            11.75
4      2017       42377              4875            11.50
5      2018       42837              4967            11.60

**Default rate by employment**

In [245]:
#Comparing the employment type and the default rate

default_by_employment = loan_dataset.groupby("EmploymentType")["Default"].agg(Total_Loans="count",Defaults="sum").reset_index()

#Calculating the default rate percentage

default_by_employment["DefaultRate"] = (default_by_employment["Defaults"] / default_by_employment["Total_Loans"] * 100).round(2)

default_by_employment.to_csv("employment_default_rate_summary.csv", index=False)

In [246]:
default_by_employment

EmploymentType  Total_Loans  Defaults  DefaultRate
0      Full-time        63656      6024         9.46
1      Part-time        64161      7677        11.97
2  Self-employed        63706      7302        11.46
3     Unemployed        63824      8650        13.55